### PR TITLE
[redis] Initial stubs for `redis.asyncio.cluster`

### PR DESCRIPTION
**Pull request by juanamari94**

### CODE REVIEW

**Comment by juanamari94** [comment]
Hey @AlexWaygood , I think I've figured out most of the stub errors but I'm running into the following issue:

```
error: redis.asyncio.cluster.ClusterPipeline is inconsistent, metaclass differs
Stub: at line 115 in file /home/runner/work/typeshed/typeshed/stubs/redis/redis/asyncio/cluster.pyi
N/A
Runtime: at line 1308 in file /tmp/tmpb89zcsbw/lib/python3.10/site-packages/redis/asyncio/cluster.py
<class 'typing._ProtocolMeta'>

error: redis.asyncio.cluster.RedisCluster is inconsistent, metaclass differs
Stub: at line 20 in file /home/runner/work/typeshed/typeshed/stubs/redis/redis/asyncio/cluster.pyi
N/A
Runtime: at line 85 in file /tmp/tmpb89zcsbw/lib/python3.10/site-packages/redis/asyncio/cluster.py
<class 'typing._ProtocolMeta'>
```

Have you seen it before in the past? How can I go about tackling it?

I think this might be related to `RedisCluster` and `ClusterPipeline` not properly inheriting from `AsyncRedisClusterCommands`.

**Comment by juanamari94** [comment]
@sobolevn Any idea how to get past this typing error? 

```
error: redis.asyncio.cluster.ClusterPipeline is inconsistent, metaclass differs
Stub: at line 115 in file /home/runner/work/typeshed/typeshed/stubs/redis/redis/asyncio/cluster.pyi
N/A
Runtime: at line 1308 in file /tmp/tmp6zn4pd2z/lib/python3.10/site-packages/redis/asyncio/cluster.py
<class 'typing._ProtocolMeta'>
```

**Comment by juanamari94** [comment]
Bumping this, @AlexWaygood don't mean to bother at all 😓 but I'm a bit flabbergasted at the stub test error and would really appreciate some help trying to trouble shoot it so we can get these changes out the door and have more complete types for Redis

**Comment by AlexWaygood** [comment]
Sorry for the delayed response @juanamari94 😞

I'll try to help out with this today or tomorrow!

**Comment by srittau** [comment]
@juanamari94 I assume the error you are seeing is because the implementation derives from `AsyncRedisClusterCommands`, which probably has a metaclass in it's bases. I would suggest to either add the class, including its base classes, but keep the body empty for now, or to `# type: ignore` the error for now.

**Comment by juanamari94** [comment]
Hi @srittau , thanks for the tip! I'll give it a shot.

**Comment by srittau** [comment]
@juanamari94 Correction: If you add the class, add `def __getattr__(self, name: str) -> Incomplete: ...` to the body so stubtest doesn't complain about the class not matching the implementation.

**Comment by srittau** [comment]
Sorry, my bad. `# type: ignore` was nonsense. You can add the lines `redis.asyncio.cluster.ClusterPipeline` and `redis.asyncio.cluster.RedisCluster` to `stubs/redis/@tests/stubtest_allowlist.txt` (with a short comment). This is stubtest error, not a mypy error.

**Comment by juanamari94** [comment]
@srittau Good idea! It seems that there's already a small comment regarding metaclass differing and also certain #todos. Hope this will be enough!

**Comment by juanamari94** [comment]
@srittau Seems that other CI tests are now failing, potentially due to the value missing from their `allowlist.txt`, should I also add the lines `redis.asyncio.cluster.ClusterPipeline` and `redis.asyncio.cluster.RedisCluster` as corresponding entries to each of those allowlists?

File: stubs/redis/redis/asyncio/cluster.pyi [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    def get_default_node(self) -> ClusterNode: ...
+    def set_default_node(self, node: ClusterNode) -> None: ...
+    def get_node(self, host: str | None = ..., port: int | None = ..., node_name: str | None = ...) -> ClusterNode | None: ...
+    def get_node_from_key(self, key: str, replica: bool = ...) -> ClusterNode | None: ...
+    def keyslot(self, key: EncodableT) -> int: ...
+    def get_encoder(self) -> Encoder: ...
+    def get_connection_kwargs(self) -> dict[str, Any | None]: ...
+    def set_response_callback(self, command: str, callback: ResponseCallbackT) -> None: ...
+    async def execute_command(self, *args: EncodableT, **kwargs: Any) -> Any: ...
+    def pipeline(self, transaction: Any | None = ..., shard_hint: Any | None = ...) -> ClusterPipeline: ...
```
Comment (srittau) [inline stubs/redis/redis/asyncio/cluster.pyi:81]:
```suggestion
    def pipeline(self, transaction: Any | None = ..., shard_hint: Any | None = ...) -> ClusterPipeline[_StrType]: ...
```

**Comment by srittau** [comment]
I'm not sure why mypy_primer fails currently, but it's most likely unrelated to this PR.

**Comment by juanamari94** [comment]
@AlexWaygood yay

**Comment by juanamari94** [comment]
@srittau All checks pass now! Really appreciate your help!

**Comment by juanamari94** [comment]
Hey @AlexWaygood , how are you? Just wanted to follow up and ask if there's anything else that might need doing in this pull request!

**Review by AlexWaygood — COMMENTED** [review]
Looks good, but it's missing quite a few attribute annotations. I'll add them and then merge.

**Review by AlexWaygood — APPROVED** [review]
Thanks for the PR, and sorry for the delay in getting to this! It's been busy round here recently :)

**Comment by juanamari94** [comment]
All good! Happy to have contributed, thank you for your patience, too!

**Pull request merged**